full-service/src/db/txo.rs

```diff
@@ -639,23 +635,52 @@ impl TxoModel for Txo {
     }
 
     fn list_unspent(
-        account_id_hex: &str,
+        account_id_hex: Option<&str>,
         assigned_subaddress_b58: Option<&str>,
         token_id: Option<u64>,
         offset: Option<u64>,
         limit: Option<u64>,
         conn: &Conn,
     ) -> Result<Vec<Txo>, WalletDbError> {
-        use crate::db::schema::txos;
+        use crate::db::schema::{transaction_input_txos, transaction_logs, txos};
+
+        /*
+            SELECT * FROM txos
+            LEFT JOIN transaction_txos
+            ON txos.id = transaction_txos.txo_id
+            LEFT JOIN transaction_logs
+            ON transaction_txos.transaction_log_id = transaction_logs.id
+            WHERE (transaction_logs.id IS NULL
+            OR ((transaction_txos.used_as = "input" AND (transaction_logs.failed = 1 OR transaction_logs.submitted_block_index = null))
+            OR (transaction_txos.used_as != "input" AND transaction_logs.failed = 0)))
+            AND txos.key_image IS NOT NULL
+            AND txos.spent_block_index IS NULL
+        */
+
+        let mut query = txos::table
+            .into_boxed()
+            .left_join(transaction_input_txos::table)
+            .left_join(
+                transaction_logs::table
+                    .on(transaction_logs::id.eq(transaction_input_txos::transaction_log_id)),
+            );
 
-        let mut query = txos::table.into_boxed();
+        if let Some(account_id_hex) = account_id_hex {
+            query = query.filter(txos::account_id_hex.eq(account_id_hex));
+        }
 
-        query = query
-            .filter(txos::received_account_id_hex.eq(account_id_hex))
-            .filter(txos::subaddress_index.is_not_null())
-            .filter(txos::key_image.is_not_null())
-            .filter(txos::pending_tombstone_block_index.is_null())
-            .filter(txos::spent_block_index.is_null());
+        query = query.filter(
+            transaction_logs::id
+                .is_null()
+                .or(transaction_logs::failed.eq(true))
+                .or(transaction_logs::id
+                    .is_not_null()
+                    .and(transaction_logs::submitted_block_index.is_null())),
+        );
+
+        query = query.filter(txos::received_block_index.is_not_null());
+        query = query.filter(txos::key_image.is_not_null());
```

Comment:
I think you need to add a filter for `txos::subaddress_index.is_not_null()` here

Reply:
that's guaranteed by the key image not being null.

Reply:
wait, then what was the point of this PR?

https://github.com/mobilecoinofficial/full-service/commit/f4f807ba7429521c8a3d021bf37a08eeb3a9a4b7

Reply:
It was checking the wrong thing, it needed to check for key_image, not subaddress_index.

Reply:
checking key_image or both key_image and subaddress index results in the same thing, but checking just subaddress_index will result in both unspent AND unverified txos

Reply:
ahhhh gotchya. Thanks